<a href="https://colab.research.google.com/github/unt-iialab/INFO5731_Spring2020/blob/master/Assignments/INFO5731_Assignment_One.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **INFO5731 Assignment - 1**

This assignment includes range of programming tasks and data analysis exercises, aiming to enhance your Python skills and proficiency in string manipulation, numerical computations, and data science. Follow the instructions carefully and design well-structured Python programs to address each question.


1.  Question 1-5 using AI assistance is strictly prohibted
2.  Question 6-7 AI assitance is recommended by following guidelines.

Guidelines and expectations file can be found in week 2 module.

Late Submission will have a penalty of 10% reduction for each day after the deadline.

# Question 1 (10 points).

The formula that describes the volume V of a hemisphere with radius r is the following:

$ V=\frac{2}{3}\ast\ \pi\ast\ r^3 $

Write a program to calculate the value of V when r is in the range of even numbers from 1-20. Output the result in the following format:

	r	V
	2	…
	4	…
	6	…
	.
	.
	.
	20	…


In [1]:
x = 2/3 #assigned x value as 2/3 according to formulae
p = 22/7 #assigned pi value  to y

print("r", "V")
for i in range(1, 21):  #using for loop in range of 1 to 21
  if i%2 == 0:          #verfying the integer is even, if even integer goes to next step
    r = i**3            #calaculating the cube of radius
    V = x*p*r           # multilplying the all values to get Volume
    print(i, V)


r V
2 16.76190476190476
4 134.09523809523807
6 452.5714285714285
8 1072.7619047619046
10 2095.238095238095
12 3620.571428571428
14 5749.333333333332
16 8582.095238095237
18 12219.428571428569
20 16761.90476190476


# Question 2 (10 points)

(a) Write a Python program to generate a sequence of 12 random numbers falling within the range of 50 to 150.

(b) Arrange the generated numbers in ascending order. Display the sorted list.

(c) Extend the program to arrange the numbers in descending order as well. Present the sorted list accordingly.

(d) Compute and output the average of the entire group of generated numbers.

(e) Enhance the program to calculate and display the average of the first 6 numbers from both the ascending and descending sorted lists.

In [2]:
# Write code for each of the sub parts with proper comments.
import numpy as np                                              #imported numpy library to generate random numbers

x = np.random.randint(50, 151, size=12)                        #By using random in range of 50 to 151 with size of 12 generated 12 random numbers
print("Sequence random numbers:",x)

x.sort()                                                       #By using sort() function sorted the random numbers in permanently in ascending order
print("Ascending Order:",x)

y = x[::-1]                                                    #By slicing list from last to first with negative index to get descending order
print("Descending Order:",y)

av = sum(x)/12
print("Avergae of entier list is :", av)                        #calculated the average of list by sum function and total size of list

fisrt6_asc_avg = sum(x[6:])/6                                   #calculated the average of first 6 numbers of ascending list by slicing the list in sum function
print("Average of Fisrt 6 ascending numbers:",fisrt6_asc_avg)

first6_dsc_avg = sum(y[6:])/6                                   #calculated the average of first 6 numbers of descending list by slicing the list in sum function
print("Average of Fisrt 6 descending numbers:",first6_dsc_avg)

Sequence random numbers: [106 117 119 117 127  88  73 143  58  78  66 101]
Ascending Order: [ 58  66  73  78  88 101 106 117 117 119 127 143]
Descending Order: [143 127 119 117 117 106 101  88  78  73  66  58]
Avergae of entier list is : 99.41666666666667
Average of Fisrt 6 ascending numbers: 121.5
Average of Fisrt 6 descending numbers: 77.33333333333333


# Question 3 (10 points).

Create a Python program that encompasses various string manipulation task:

(a) Prompt the user to input two sentences.

(b) Transform each sentence into a list, where each element corresponds to a word in the sentence. For example, if the sentences are "Hello, how are you doing today?" and "This is an example sentence for the program!"

The program should output:
['Hello', ',', 'how', 'are', 'you', 'doing', 'today', '?', 'This', 'is', 'an', 'example', 'sentence', 'for', 'the', 'program', '!']

(c) Compare the two sentences. Print out a list of words that occur in both sentences.

(d) Remove any punctuation from the sentence and print the sanitized sentences. It should remove at least the following characters, but it can remove more: period(“.”), comma (“,”), semicolon (“;”), and colon (“:”).

(e) Count the number of vowels in each of the sentences. Print out each vowel and the occurrences  in the each of the two sentences separately, such as:

Sentence1: a:2, e:2, i:0, etc.
Sentence2: a:1, e:2, i:2, etc.

Design a well-structured Python program to execute these tasks efficiently.

In [21]:
import string


#Part-A
input1 = input("Enter the sentence1: ")  #inputing the 1st sentence
input2 = input("Enter the sentence2: ")  #inputing the 2nd sentence

#Part-B removing punctuation

a = input1.split()  # spilting the each word of input 1
b = input2.split()  # spilting the each word in input2 sentence

sentence1 = [] # intiating the empty sentence lists two add each word of sentence
sentence2 = []
pun = string.punctuation   # intaiting the pun to all puncations stoer in string.puncations method

for i in a:   #using for loop

    if i[-1] in pun:   # verfying the each word in puncations in neagtive index
      sentence1.append(i[:-1])  # stroing the puncations in sentence1 list
      sentence1.append(i[-1])
    else:
      sentence1.append(i) # if word doesnt have the puncation directly it will append to sentence 1 list

for i in b:     # same as above loop

    if i[-1] in pun:
      sentence2.append(i[:-1])
      sentence2.append(i[-1])
    else:
      sentence2.append(i)
print("Sanitize Sentences: ")
print(sentence1) # print the result
print(sentence2)

#Part-C Compare the simsilar words in sentences

compare_list = []  # intating the empty list

for i in sentence1:   # in this loop coparing the each word in sentence,  right now taking the 1st word in sentence1[0] list
  for j in sentence2: # in this innner loop taking the 1st word in sentece2[0] loop,  after completing the sentece2[0] it will the verfy upto sentence2[n](last index word in sentence2) after completing the comlete len of sentnece 2 it verfys the sentence[1]
    if i == j:   # comparing the sentence1[0] and senrecne[0] words
      compare_list.append(i) # if it finds similar  append to compare list

print("Similar Words: ",compare_list)


#Part-D Remove ,:;.

santized_sentence1 = ''  # intaiting the new sentence1 string
santized_sentence2 = ''  # intaiting the new sentence2 string

for i in input1:  # using for loop, taking the input1 first character
  if i != '.' and i != ',' and i != ';' and i != ':':   # verfying  char is not falls in conditions
    santized_sentence1 += i   # if char not falls in condition it  update into santized_sentence1 string

for i in input2:
  if i != '.' and i != ',' and i != ';' and i != ':':
    santized_sentence2 += i


print(santized_sentence1)
print(santized_sentence2)


#Part-E couunt the vowles
input1 = input1.lower()  #converting the input1 sentence into lower case for dictionary to count the vowels
input2 = input2.lower()  #converting the input2 sentence into lower case for dictionary to count the vowels

vowels = 'aeiou' # intiating vowels to vowels

dict1 = {'a': 0, 'e': 0, 'i': 0, 'o': 0, 'u': 0} # intiaiting the dictonary(dict1) with null count of each vowel
dict2 = {'a': 0, 'e': 0, 'i': 0, 'o': 0, 'u': 0} # intiaiting the dictonary(dict2) with null count of each vowel

for i in input1:  # using for loop, to count vowels veryfing the each char input1(0)....input(n)
  if i in vowels:  # verfying the 1st char of falls in vowles
    dict1[i] += 1  # if char in vowles it will upadte the count into dict1

for i in input2:
  if i in vowels:
    dict2[i] += 1

print("Sentenc1: ")
for i , j in dict1.items():   #using the for loop, pirnting the each item in dict as i and j
  if j > 0: # if any count more than zero results are printing
    print(i, ":", j)

print("Sentenc2: ")
for i , j in dict2.items():
  if j > 0:
    print(i, ":", j)

#Hello, how are you doing today?
#This is an example sentence for the program!

Enter the sentence1: Hello, how are you doing today?
Enter the sentence2: This is an example sentence for the program!
['Hello', ',', 'how', 'are', 'you', 'doing', 'today', '?']
['This', 'is', 'an', 'example', 'sentence', 'for', 'the', 'program', '!']
Similar Words:  []
Hello how are you doing today?
This is an example sentence for the program!
Sentenc1: 
a : 2
e : 2
i : 1
o : 5
u : 1
Sentenc2: 
a : 3
e : 6
i : 2
o : 2


# Question 4 (10 Points)

Design and implement a simple command-line calculator program. The program should handle both integers and floating-point numbers.

(a) Write program that takes a mathematical expression as input and evaluates it. The program should support addition, subtraction, multiplication, and division.

(b) Extend the calculator to handle parentheses in mathematical expressions. Implement the necessary logic to evaluate expressions with nested parentheses.

(c) Enhance your calculator to include additional mathematical operations such as exponentiation, square root, and modulus.

(d) Implement error handling in the calculator program. Handle cases such as division by zero and invalid input expressions. Provide meaningful error messages.

(e) Implement a command to exit the calculator.

In [102]:
import math
import re

def calculator(E): # calculator Function is used to take math expresion and evaluates it and returns the value
    try:
        E = E.replace('^', '**')


        # Function to evaluate expressions with parentheses
        def evaluate(tokens):
            def apply_operator(operators, values):
                right = values.pop()
                left = values.pop() if values else 0  # Handle unary operations
                op = operators.pop()
                if op == '+':
                    values.append(left + right)
                elif op == '-':
                    values.append(left - right)
                elif op == '*':
                    values.append(left * right)
                elif op == '/':
                    if right == 0:
                        raise ZeroDivisionError("Division by zero is not allowed.")
                    values.append(left // right)  # Integer division
                elif op == '**':
                    values.append(left ** right)
                elif op == '%':
                    values.append(left % right)

            # Define operator precedence
            precedence = {'+': 1, '-': 1, '': 2, '/': 2, '*': 3, '%': 2}
            values = [] # Stack for numbers
            operators = [] # Stack for operator
            i = 0 # Token index

            while i < len(tokens):
                token = tokens[i]
                # Handle numbers
                if isinstance(token, str) and token.isdigit():
                    values.append(int(token))
                elif token == 'sqrt': # Handle square root function
                    if i + 1 < len(tokens) and tokens[i + 1] == '(':
                        operators.append(token)
                    else:
                        return "Error: sqrt function requires parentheses."
                elif token in precedence: # Handle arithmetic operators
                    while (operators and operators[-1] in precedence and
                           precedence[operators[-1]] >= precedence[token]):
                        apply_operator(operators, values)
                    operators.append(token)
                elif token == '(':
                    operators.append(token)
                elif token == ')':
                    while operators and operators[-1] != '(':
                        apply_operator(operators, values)
                    operators.pop()  # Remove '('
                    if operators and operators[-1] == 'sqrt':
                        operators.pop()
                        values.append(int(math.sqrt(values.pop())))
                i += 1

            while operators:
                apply_operator(operators, values)

            return values[0] if values else "Error: Invalid expression."

        tokens = re.findall(r'sqrt|\d+|[()+\-*/%^]', E) # Tokenize the input expression
        return evaluate(tokens)
    except ZeroDivisionError:
        return "Error: Dividing by Zero is not possible. Try again"
    except (SyntaxError, NameError, TypeError):
        return "Error: Thats not a valid input. Please enter a valid mathematical expression."

def main():
    print("Hello!! Welcome to karthik's Command line Calculator!")
    print("Valid Operations: +, -, *, /, ^, %, sqrt()")
    print("type 'exit' to quit the calculator.")

    while True: # looping for user input
        userInput = input("Enter a math expression:")
        if userInput.lower() == 'exit':
            print("Thank you for using my calculator :)")
            break
        output = calculator(userInput)
        print("Result:", output)


if __name__ == "__main__":
    main()


Hello!! Welcome to karthik's Command line Calculator!
Valid Operations: +, -, *, /, ^, %, sqrt()
type 'exit' to quit the calculator.
Enter a math expression:exit
Thank you for using my calculator :)


# Question 5 (20 points).

In the field of Data Science, data is often formatted as a comma-delimited (CSV) file, where each line in a file corresponds to a field's value. Refer to Chapter 9 (p. 106) for more information on comma-delimited files.

To answer this question, you will need to download the file [Assignment1_denton_housing.csv](https://raw.githubusercontent.com/unt-iialab/info5731-spring2022/main/assignments/Assignment1_denton_housing.csv) included with this assignment. The file contains statistics about housing in the city of Denton from 2008 to 2014. Write a program to implement the following questions.

(a) (5 pts) Calculate the difference in the number of Occupied Housing Units from year to year and print it. The difference must be calculated for the consecutive years such as 2008-2009, 2009-2010 etc. Finally, print the values in the ascending order.

(b) (5 pts) For all the years, calculate the percentage of housing units which are vacant an occupied. Print the results in the following format:

Year Vacant Housing Units Occupied Housing Units

    2008  30%   70%                          
	2009 ----- -----
	2010 ----- -----
    2011 ----- -----
	2012 ----- -----
    2013 ----- -----
	2014 ----- -----
  
  

(c) (5 pts) Calculate and print the valued and years in which the highest number of housing units were vacant and occupied. Print the results in the following format:  

                              Year	Value
	Vacant Housing Units	  -----   -----
	Occupied Housing Units	-----   -----

(d) (5 pts) Calculate the harmonic mean of the total housing units and print it out. You can find the information about harmonic mean here: https://ncalculators.com/statistics/harmonic-mean-calculator.htm

In [27]:
import numpy as np
import pandas as pd

df = pd.read_csv("https://raw.githubusercontent.com/unt-iialab/info5731-spring2022/main/assignments/Assignment1_denton_housing.csv")

print(df)

    year             title_field  value
0   2014     Total Housing Units  49109
1   2014    Vacant Housing Units   2814
2   2014  Occupied Housing Units  46295
3   2013     Total Housing Units  47888
4   2013    Vacant Housing Units   4215
5   2013  Occupied Housing Units  43673
6   2012     Total Housing Units  45121
7   2012    Vacant Housing Units   3013
8   2012  Occupied Housing Units  42108
9   2011     Total Housing Units  44917
10  2011    Vacant Housing Units   4213
11  2011  Occupied Housing Units  40704
12  2010     Total Housing Units  44642
13  2010    Vacant Housing Units   3635
14  2010  Occupied Housing Units  41007
15  2009     Total Housing Units  39499
16  2009    Vacant Housing Units   3583
17  2009  Occupied Housing Units  35916
18  2008     Total Housing Units  41194
19  2008    Vacant Housing Units   4483
20  2008  Occupied Housing Units  36711
21  2014          Owner Occupied  21427
22  2014         Renter Occupied  24868
23  2013          Owner Occupied  20845


In [60]:
# Your code here
a = df[df['title_field'] == 'Occupied Housing Units'] # filtered out the only Occupied Housing Units data
a = a.sort_values(by='year') # sorted by year the data
a['difference'] = a['value'].diff()  # calculating the consecytive difference
a = a.dropna(subset=['difference']) # droping the null values in data

list = [] # intiatiezed the empty new list


for i in range(len(a) - 1):  # using for loop, from a data of a[0](year1, year2, diiference value) appending to new list
    y1 = a.iloc[i]['year']
    y2 = a.iloc[i + 1]['year']
    res = a.iloc[i + 1]['difference']
    list.append((y1, y2, res))


x = sorted(list, key=lambda x: x[2]) # sorting the new list in ascending order by difference values in X

for y1, y2, diff in x: # printing the result in given formart as 2009-2010 and difference
    print(f"{y1} - {y2}: {diff}")


2010 - 2011: -303.0
2011 - 2012: 1404.0
2012 - 2013: 1565.0
2013 - 2014: 2622.0
2009 - 2010: 5091.0


In [70]:
a = df[df['title_field'] == 'Vacant Housing Units'] # storing only the Vacant Housing Units data in a
b = df[df['title_field'] == 'Occupied Housing Units']  # storing only the Occupied Housing Units data in b
c = df[df['title_field'] == 'Total Housing Units']  # storing only the Occupied Housing Units data in c


res_df = a[['year', 'value']].merge(b[['year', 'value']], on='year', suffixes=('_vacant', '_occupied')) # mergeing the a nd b of Vacant Housing Units and Occupied Housing Units data
res_df = res_df.merge(c[['year', 'value']], on='year', suffixes=('', '_total'))  # merging the above merged data and Occupied Housing Units data

res_df['percentage_vacant'] = (res_df['value_vacant'] / res_df['value']) * 100. #calaculating the percentgae of Vacant Housing Units and total housing
res_df['percentage_occupied'] = (res_df['value_occupied'] / res_df['value']) * 100. #calaculating the percentgae of ccupied Housing Units and total housing


for _, row in res_df.iterrows():   #printing the output in given format
    print(f"{row['year']}  {row['percentage_vacant']:.2f}  {row['percentage_occupied']:.2f}")


2014.0  5.73  94.27
2013.0  8.80  91.20
2012.0  6.68  93.32
2011.0  9.38  90.62
2010.0  8.14  91.86
2009.0  9.07  90.93
2008.0  10.88  89.12


In [91]:
a = df[df['title_field'] == 'Vacant Housing Units'] # storing only the Vacant Housing Units data in a
b = df[df['title_field'] == 'Occupied Housing Units']  # storing only the Occupied Housing Units data in b

a = a.loc[a['value'].idxmax()] # using idmax method to get the maximum value of Vacant Housing Units
b = b.loc[b['value'].idxmax()] # using idmax methos to get the maximum values of Occupied Housing Units

# Print the results in the desired format
print(f"                       Year Value ")
print(f"Vacant Housing Units   {a['year']} {a['value']} ")
print(f"Occupied Housing Units {b['year']} {b['value']} ")

                       Year Value 
Vacant Housing Units   2008 4483 
Occupied Housing Units 2014 46295 


In [94]:
c = df[df['title_field'] == 'Total Housing Units']  # storing only the Occupied Housing Units data in c
values = c['value'] # storing the only totoal housing units of values into values
i = len(values) #calculating the len of values
hm = i / (1 / values).sum() # harmonic formaulae len(values)/(1/ sum of values)
print('Harmonic Mean of Total Housing Units: ',hm)


Harmonic Mean of Total Housing Units:  44398.28543969185


# **For question 6 and 7 you can take AI assitance to solve.**

Make sure you follow the instructions given in class to answer these questions.


**Important points:**

1.   You must come up with at least 10 edge cases.
2.   You should provide prompts and revised prompts given to ChatGPT.
2.   Debug your code and write comments for each block of code
3.   Analyzing error will not consider as edge case.

Edge cases are like real life scienarios which are realated to the given question.

# Question 6 (20 points).
Develop a advance Python program to manage inventory records for a prodcut managing company. The system needs functions for adding, deleting and changing products and basic search options. Your system should use CSV files to automate restocking and produce reports that show item values alongside category breakdowns. Manage edge cases and corrupted CSV files. Finally maintaine 50 products of different categories in inventory.

# Question 7 (20 points)

Create a well desined Python program for a flight booking system. Offer core functions for adding new flights, finding trips, reserving and releasing seating space and creating performance reports. Ensure the system manages edge cases like invalid input, overbooking, and cancellations. Finally, extend the booking system by including advanced features and maintain about 50 flight in the system.

In [ ]:
# Your code here


# Mandatory Question

Provide your thoughts on the assignment. What did you find challenging, and what aspects did you enjoy? Your opinion on the provided time to complete the assignment. Share any suggestions for improvement. Your feedback will be valuable for enhancing the learning experience.

In [ ]:
# Write your response below